pkg/controller/apicuriodeployment/apicuriodeployment_controller.go

```diff
@@ -0,0 +1,279 @@
+}
+
+func (r *ReconcileApiCurioDeployment) bootstrap(request reconcile.Request) error {
+	var err error
+
+	if r.tmpl == nil {
+		restCfg := r.config
+		namespace := request.Namespace
+		r.tmpl, err = openshift.NewTemplate(namespace, restCfg, openshift.TemplateDefaultOpts)
+
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileApiCurioDeployment) processTemplate(cr *integreatlyv1alpha1.ApicurioDeployment) ([]runtime.RawExtension, error) {
+	baseTmplPath := "/etc/apicurio-operator/res"
+	tmplPath := cr.Spec.Template
+	if tmplPath == "" {
+		return nil, fmt.Errorf("Spec.Template.Path property is not defined")
+	}
+
+
+	res, err := openshift.LoadKubernetesResourceFromFile(baseTmplPath + "/" + tmplPath)
+	if err != nil {
+		return nil, err
+	}
+
+	params := make(map[string]string)
+	for k, v := range routeParams {
+		if k == "AUTH_ROUTE" && cr.Spec.ExternalAuthUrl != "" {
+			params[k] = cr.Spec.ExternalAuthUrl
+			continue
+		}
+		params[k] = v + "." + cr.Spec.AppDomain
+	}
+
+	if cr.Spec.AuthRealm != "" {
+		params["KC_REALM"] = cr.Spec.AuthRealm
+	}
+
+	if len(cr.Spec.JvmHeap) == 2 {
+		params["API_JVM_MIN"] = cr.Spec.JvmHeap[0]
+		params["API_JVM_MAX"] = cr.Spec.JvmHeap[1]
+	}
+	if len(cr.Spec.MemLimit) == 2 {
+		params["API_MEM_REQUEST"] = cr.Spec.MemLimit[0]
+		params["API_MEM_LIMIT"] = cr.Spec.MemLimit[1]
+	}
+
+	tmpl := res.(*v1template.Template)
+	r.tmpl.FillParams(tmpl, params)
+	ext, err := r.tmpl.Process(tmpl, params, openshift.TemplateDefaultOpts)
+
+	return ext, err
+
+}
+
+func (r *ReconcileApiCurioDeployment) getRuntimeObjects(exts []runtime.RawExtension) ([]runtime.Object, error) {
+	objects := make([]runtime.Object, 0)
+	for _, ext := range exts {
+		res, err := openshift.LoadKubernetesResource(ext.Raw)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, res)
+	}
+
+	return objects, nil
+}
+
+func (r *ReconcileApiCurioDeployment) transformObject(o runtime.Object) (*unstructured.Unstructured, error) {
+	uo, err := openshift.UnstructuredFromRuntimeObject(o)
+	if err != nil {
+		return nil, err
+	}
+
+	return uo, nil
+}
+
+func (r *ReconcileApiCurioDeployment) createObject(o runtime.Object, cr *integreatlyv1alpha1.ApicurioDeployment) error {
+	err := r.client.Create(context.TODO(), o)
+	if err != nil && !kuberr.IsAlreadyExists(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileApiCurioDeployment) createObjects(objects []runtime.Object, ns string, cr *integreatlyv1alpha1.ApicurioDeployment) error {
+	for _, o := range objects {
+		uo, err := r.transformObject(o)
+		if err != nil {
+			return fmt.Errorf("failed to transform object: %v", err)
+		}
+
+		uo.SetNamespace(ns)
+		err = controllerutil.SetControllerReference(cr, uo, r.scheme)
+		if err != nil {
+			return fmt.Errorf("failed to set owner in object: %v", err)
+		}
+
+		//ignore auth objects if using external keycloak
+		isAuthObj := strings.Contains(uo.GetName(), "auth")
+		if cr.Spec.ExternalAuthUrl != "" && isAuthObj {
+			continue
```

Comment:
It will skip all keycloak related runtime objects if `external_auth_url` is defined in the cr.

Reply:
Is this custom to apicuro? How would something like this be catered for with the lib?

Reply:
Yes it is, I think it can be dealt with outside of the process function (we can also create a pre-processing function to be used by the method caller).